docs/proposed/pub-sub.rst

```diff
@@ -0,0 +1,203 @@
+
+Magic Folders in Tahoe currently poll, looking for updates. This
+involves periodically downloading some mutable directory-capabilities
+and determining if anything has changed (wasting bandwidth). It would
+be nice if a client, Alice, could tell other participants in the
+magic-folder that she's added some content. More generally, it would
+be useful for a client to be able to ask a storage-server to tell it
+whenever a particular (mutable) capability has changed. (For the Magic
+Folders use-case, this would mean Alice simply makes changes as now,
+and relies on the server to update interested clients).
+
+
+Grid -> Client Communication
+````````````````````````````
+
+Allowing the operator(s) of a Grid to communicate with clients is
+often desired. For example, notifying users of upcoming maintenance,
+new terms or general news. There is currently no mechanism to do this.
+
+An obvious way to implement this right now would be with a mechanism
+like Magic Folders uses: a mutable capability containing news which
+clients poll (and which the operators can update). This would suffer
+the same polling problems described above.
+
+
+Some Required Definitions
+=========================
+
+Slot -- unit of storage for SDMF and MDMF mutable files, referenced by
+        a Storage Index.
+
+Storage Index -- first 16 bytes of SHA256d("allmydata_mutable_readkey_to_storage_index_v1" + Read Key)
+
+Read Key -- first 16 bytes of SHA256d("allmydata_mutable_writekey_to_readkey_v1" + Write Key)
+
+Write Key -- first 16 bytes of SH256d("allmydata_mutable_privkey_to_writekey_v1" + Signature Key)
+
+Signature Key -- private part of SDMF keypair
+
+Verification Key -- public part of SDMF keypair
+
+SDMF keypair -- random public/private keypair (RSA 2048-bit)
+
+
+Proposed Solution
+=================
+
+Anything wanting a "broadcast" or "pub-sub" like pattern can implement
+it using the following mechanism:
+
+ - the publisher creates a mutable directory, yielding a write-cap, "W"
+ - anyone wishing to subscribe is given a read-cap, "R"
+   - anyone possessing a valid read-cap "R" (including the publisher)
+     can obviously "invite" further clients
+   - nobody else may subscribe
+ - to add information, the publisher adds a new file to the mutable
+   directory
+ - old information can be removed (or not) as the publisher wishes
+
+Without further changes to Tahoe-LAFS, this can already be achieved:
+clients can "poll" the read-capability and determine if anything new
+exists.
+
+
+Improvements Over Polling
+-------------------------
+
+The above mechanism would scale somewhat poorly as the number of
+"broadcast" capabilities increased. It would be better if the
+storage-server could *tell* clients when a mutable Slot has changed.
+
+Storage servers implement a new feature, based on WebSockets. See also
+:ref:`http-storage-node-protocol` with which this protocol aims to be
+compatible. A single WebSocket endpoint exists:
+
+ - <server>/v1/mutable_updates
```

Comment:
fwiw, to get in on the bike shedding early, I like `-` as a word separator in URLs rather than `_`.

Reply:
okay, that's fine with me :)